injectproxy/enforce_test.go

```diff
@@ -357,3 +366,533 @@ func TestEnforce(t *testing.T) {
+					`up{job!="foo",job="bar"}`,
+					false,
+				},
+				{
+					`job="fred"`,
+					`up{job!="foo",job="fred"}`,
+					false,
+				},
+
+				// Not equal label selector in the expression.
+				{
+					`job!=""`,
+					`up{job!="",job!="foo"}`,
+					false,
+				},
+				{
+					`job!="foo"`,
+					`up{job!="foo"}`,
+					false,
+				},
+				{
+					`job!="bar"`,
+					`up{job!="bar",job!="foo"}`,
+					false,
+				},
+				{
+					`job!="fred"`,
+					`up{job!="foo",job!="fred"}`,
+					false,
+				},
+
+				// Regexp label selector in the expression.
+				{
+					`job=~""`,
+					`up{job!="foo",job=~""}`,
+					false,
+				},
+				{
+					// up{job!="foo",job=~"foo"} would return no result.
+					`job=~"foo"`,
+					``,
+					true,
+				},
+				{
+					`job=~"bar"`,
+					`up{job!="foo",job=~"bar"}`,
+					false,
+				},
+				{
+					`job=~"fred"`,
+					`up{job!="foo",job=~"fred"}`,
+					false,
+				},
+				{
+					`job=~"foo|fred"`,
+					`up{job!="foo",job=~"foo|fred"}`,
+					false,
+				},
+				{
+					`job=~"foo|bar"`,
+					`up{job!="foo",job=~"foo|bar"}`,
+					false,
+				},
+
+				// Not-regexp label selector in the expression.
+				{
+					`job!~""`,
+					`up{job!="foo",job!~""}`,
+					false,
+				},
+				{
+					`job!~"foo"`,
+					`up{job!="foo",job!~"foo"}`,
+					false,
+				},
+				{
+					`job!~"bar"`,
+					`up{job!="foo",job!~"bar"}`,
+					false,
+				},
+				{
+					`job!~"fred"`,
+					`up{job!="foo",job!~"fred"}`,
+					false,
+				},
+				{
+					`job!~"foo|fred"`,
+					`up{job!="foo",job!~"foo|fred"}`,
+					false,
+				},
+				{
+					`job!~"foo|bar"`,
+					`up{job!="foo",job!~"foo|bar"}`,
+					false,
+				},
+			},
+		},
+
+		// Regexp matcher enforcer.
+		{
+			enforcedMatcher: mustNewMatcher(labels.MatchRegexp, "job", "foo|bar"),
+			stcs: []subTestCase{
+				// No selector in the expression for the enforced label.
+				{
+					``,
+					`up{job=~"foo|bar"}`,
+					false,
+				},
+
+				// Equal label selector in the expression.
+				{
+					`job=""`,
+					``,
+					true,
+				},
+				{
+					`job="foo"`,
+					`up{job="foo",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job="bar"`,
+					`up{job="bar",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job="fred"`,
+					``,
+					true,
+				},
+
+				// Not equal label selector in the expression.
+				{
+					`job!=""`,
+					`up{job!="",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job!="foo"`,
+					`up{job!="foo",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job!="bar"`,
+					`up{job!="bar",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job!="fred"`,
+					`up{job!="fred",job=~"foo|bar"}`,
+					false,
+				},
+
+				// Regexp label selector in the expression.
+				{
+					`job=~""`,
+					``,
+					true,
+				},
+				{
+					`job=~"foo"`,
+					`up{job=~"foo",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job=~"bar"`,
+					`up{job=~"bar",job=~"foo|bar"}`,
+					false,
+				},
+				{
+					`job=~"fred"`,
```

Comment:
these are disjoint, no? so why don't we error here?

Reply:
good catch!